### PR TITLE
Put along axis + fixe for partition grad

### LINKED ISSUE #1428: [BUG] Gradients for mx.topk
**Describe the bug**
Gradients for mx.core.topk seem to be incorrectly computed.

**To Reproduce**

```python
import mlx.core as mx
import mlx.nn as nn
import numpy as np
import torch

a = [[1,2,6,4,5],[9,5,6,7,8]]
x = mx.array(a, mx.float32)
xt = torch.tensor(a, dtype=torch.float32, requires_grad=True)
out_mx = mx.topk(x, 2)

out_t = torch.topk(xt, 2)[0]
loss_t = out_t.sum()
loss_t.backward()

def loss_topk(x, k):
    return mx.topk(x, k).sum()

loss_mx, grads_mx = mx.value_and_grad(loss_topk, argnums=(0))(x, 2)
print("X:")
print(x)
print("\nOutput MLX: ")
print(out_mx)
print("\nOutput Pytorch: ")
print(out_t)
print("\nLoss MLX: ")
print(loss_mx)
print("\nLoss Pytorch: ")
print(loss_t)
print("\nGrads MLX: ")
print(grads_mx)
print("\nGrads Pytorch: ")
print(xt.grad)
```

**Output**
```
X:
array([[1, 2, 6, 4, 5],
       [9, 5, 6, 7, 8]], dtype=float32)

Output MLX:
array([[5, 6],
       [8, 9]], dtype=float32)

Output Pytorch:
tensor([[6., 5.],
        [9., 8.]], grad_fn=<TopkBackward0>)

Loss MLX:
array(28, dtype=float32)

Loss Pytorch:
tensor(28., grad_fn=<SumBackward0>)

Grads MLX:
array([[0, 0, 1, 1, 0],
       [0, 0, 1, 1, 0]], dtype=float32)

Grads Pytorch:
tensor([[0., 0., 1., 0., 1.],
        [1., 0., 0., 0., 1.]])
```
**Expected behavior**
PyTorch output seems to be correct.

**Additional context**
MLX 0.17.3

### PR DESCRIPTION
**Pull request by awni**
- Add a `put_along_axis`
- Use it to fix the partition VJP
- Let arg partition grad be zero (as opposed to using stopgrad everywhere). This is the same as Jax.

Closes #1428 and #1427

### CODE REVIEW

**Review by barronalex — APPROVED** [review]
Thanks for the quick fix!

**Comment by sachinraja13** [comment]
Many thanks @awni , @barronalex and the entire team. You guys are truly amazing!
